### PR TITLE
feat(email): Add support for adding additional emails

### LINKED ISSUE #210: Add db support for adding a recovery email
This issue tracks updates needed to the auth-db-mysql server to support a user adding a recovery email.

Table: 
* Email
  * normalizedEmail:  (String) (Primary Key)
  * email: (String) 
  * uid: (Binary)
  * emailCode: (Binary)
  * isVerified: (Boolean)
  * isPrimary: (Boolean)
  * createdAt (Date)
  * verifiedAt (Date)

Ref: https://docs.google.com/document/d/1DOUWRbxU8ah6AL3gUTrlYKFY42jt_czOHbu4wL4E-sc/edit#, connects with https://github.com/mozilla/fxa-features/issues/27

- [x] Add DB methods, `createEmail`, `getEmails` and `deleteEmail`
- [x] Add `mem` and `mysql` support
- [x] Update tests
- [x] Update docs

### PR DESCRIPTION
**Pull request by vbudhram**
Fixes #210, https://github.com/mozilla/fxa-features/issues/27

### CODE REVIEW

File: lib/db/mysql.js [hunk 1]
```diff
@@ -674,13 +675,13 @@ module.exports = function (log, error) {
...
   }
 
-  // Update : accounts
-  // Set    : emailVerified = true
-  // Where  : uid = $1
-  var VERIFY_EMAIL = 'CALL verifyEmail_3(?)'
+  // Update : accounts, emails
+  // Set    : emailVerified = true if email is in accounts table or isVerified = true if on email table
+  // Where  : uid = $1, emailCode = $2
+  var VERIFY_EMAIL = 'CALL verifyEmail_4(?, ?)'
```
Comment (vbudhram) [inline lib/db/mysql.js]: Not sure of the historical significance but, `verifyEmail` did not actually look and verify that the `emailCode` belonged to the user account. It simply verified the users email. Which made me wonder, what was the point of the `emailCode` originally?

File: lib/db/mysql.js [hunk 2]
```diff
@@ -752,6 +753,66 @@ module.exports = function (log, error) {
...
+        data.isVerified,
+        data.isPrimary,
+        data.verifiedAt,
+        data.createdAt
+      ]
+    )
+      .catch(function(err){
+        // Signal exception is triggered when an attempt to
+        // add an email that is in the users table already.
+        if (err.errno === ER_SIGNAL_EXCEPTION) {
```
Comment (vbudhram) [inline lib/db/mysql.js]: There is probably a better way to do this? I wanted to signal that the email being added exists already in the account table or email table.

File: lib/db/schema/patch-045-046.sql [hunk 3]
```diff
@@ -0,0 +1,188 @@
+-- Add emails table and corresponding procedures to
+-- create, get and delete them.
+
+CREATE TABLE IF NOT EXISTS emails (
+  normalizedEmail VARCHAR(255) NOT NULL PRIMARY KEY,
+  email VARCHAR(255) NOT NULL,
+  uid BINARY(16) NOT NULL,
```
Comment (vbudhram) [inline lib/db/schema/patch-045-046.sql]: @rfk What are your thoughts on adding an index to `uid` for faster lookups? This table gets queried by uid and by email.

File: lib/db/schema/patch-045-046.sql [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+
+CREATE PROCEDURE `verifyEmail_4`(
+    IN `inUid` BINARY(16),
+    IN `inEmailCode` BINARY(16)
+)
+BEGIN
+    SET @updatedCount = 0;
+
+    IF (inEmailCode IS NULL) THEN
```
Comment (vbudhram) [inline lib/db/schema/patch-045-046.sql]: Am I being overly cautious and accounting for `emailCode` not being present? I wanted to keep the previous functionality that just verified the the users account email, but I am not sure if that is actually needed.

**Comment by vbudhram** [comment]
@seanmonstar @rfk @philbooth guys mind doing a sanity check on my approach before code gets overly complicated?

File: lib/db/mysql.js (hunk 2 above)
Comment (seanmonstar) [inline lib/db/mysql.js]: There should already be logic for this: https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/mysql.js#L914

File: lib/db/schema/patch-045-046.sql (hunk 4 above)
Comment (seanmonstar) [inline lib/db/schema/patch-045-046.sql]: I'd probably rather there weren't 2 possible paths here. Either we decide it's worth having a code to compare against (which it probably is), and thus require it, or we never want one.

File: lib/db/schema/patch-045-046.sql [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        e.verifiedAt,
+        e.createdAt
+    FROM
+        emails e
+    WHERE
+        uid = LOWER(inUid));
+END;
+
+CREATE PROCEDURE `deleteEmail_1` (
+    IN `inNormalizedEmail` VARCHAR(255)
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: The caller of this knows both the `uid` and the `email`, and I wonder if it's worth passing both of them in here as a sanity-check.  The query below would become `ELETE FROM emails WHERE normalizedEmail = inNormalizedEmail AND uid = inUid` and it would be an extra line of defense against accidentally deleting an email that doesn't belong to that user.

File: lib/db/schema/patch-045-046.sql [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    IF @isPrimary > 0 THEN
+      SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Can not delete a primary email address.';
+    END IF;
+
+	SELECT COUNT(*) INTO @isPrimary FROM emails WHERE normalizedEmail = inNormalizedEmail AND isPrimary = 1;
+    IF @isPrimary > 0 THEN
+      SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Can not delete a primary email address.';
+    END IF;
+
+	DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail;
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: If we assume that users will *usually* be doing the right thing, then you might be able to make this more efficient in the common case by using an "ask forgiveness, not permission" style:

* Do a `DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail AND NOT isPrimary`
* Check how many rows were deleted.  If zero rows were deleted then:
  * Check whether it exists with `isPrimary=1`, error out if so
  * Check whether it exists in the main `accounts` table, error out if so
  * Otherwise fall through, reporting that zero rows were deleted

You could also move that "check what went wrong" logic out of the stored procedure and into the javascript method if it would simplify the code some.

File: lib/db/schema/patch-045-046.sql [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+    COMMIT;
+END;
+
+DROP procedure IF EXISTS `deleteAccount_12`;
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: Why is this necessary?

File: lib/db/schema/patch-045-046.sql (hunk 4 above)
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: +1, we will thank ourselves for simplifying down to a single path here.

Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: Oh, but, the auth-server currently calls this `verifyEmail` procedure when you perform a sign-in confirmation on an unverified account, in which case it won't know the correct `emailCode` to pass in....

File: fxa-auth-db-server/index.js [hunk 8]
```diff
@@ -69,10 +69,23 @@ function createServer(db) {
...
+      return db.verifyEmail(req.params.id, req.params.emailCode)
+    })
+  )
   api.post('/account/:id/locale', withIdAndBody(db.updateLocale))
   api.get('/account/:id/sessions', withIdAndBody(db.sessions))
 
+  api.get('/account/:id/emails', withIdAndBody(db.accountEmails))
+  api.put('/account/:id/emails', withIdAndBody(db.createEmail))
+  api.post('/account/:id/emails', withIdAndBody(db.createEmail))
+  api.del('/account/:id/emails/:email',
```
Comment (philbooth) [inline fxa-auth-db-server/index.js]: These new endpoints all need documenting in `fxa-auth-db-server/docs/API.md`.

File: lib/db/mysql.js [hunk 9]
```diff
@@ -752,6 +753,66 @@ module.exports = function (log, error) {
...
+      [
+        uid
+      ]
+    )
+  }
+
+  // Delete : emails
+  // Values : email = $1
+  var DELETE_EMAIL = 'CALL deleteEmail_1(?)'
+  MySql.prototype.deleteEmail = function (email) {
```
Comment (philbooth) [inline lib/db/mysql.js]: These new methods all need documenting in `docs/API.md`.

File: lib/db/schema/patch-045-046.sql [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    DECLARE EXIT HANDLER FOR SQLEXCEPTION
+        BEGIN
+        ROLLBACK;
+        RESIGNAL;
+    END;
+
+    START TRANSACTION;
+
+    -- Don't delete any email that are considered primary emails
+	SET @isPrimary = 0;
```
Comment (philbooth) [inline lib/db/schema/patch-045-046.sql]: Looks like there's some mixed tabs and spaces messing with the indentation in here.

File: test/local/mysql_tests.js [hunk 11]
```diff
@@ -378,6 +378,87 @@ DB.connect(config)
...
+                t.equal(result.length, 1, 'one email returned')
+
+                // Verify first email is email from accounts table
+                t.equal(result[0].email, account.email, 'matches account email')
+                t.equal(!!result[0].isPrimary, true, 'isPrimary is true on account email')
+                t.equal(!!result[0].isVerified, account.emailVerified, 'matches account emailVerified')
+              }
+            )
+        }
+      )
```
Comment (philbooth) [inline test/local/mysql_tests.js]: I think we should add some tests of the potential error paths in here maybe.

File: lib/db/mysql.js (hunk 1 above)
Comment (vbudhram) [inline lib/db/mysql.js]: > I think it'd be worth changing that so that the auth-server passes in the emailCode in all cases.

Sounds good. Doing this will help simply some of the logic in the auth server.

File: lib/db/mysql.js (hunk 2 above)
Comment (vbudhram) [inline lib/db/mysql.js]: I believe `ER_DUP_ENTRY` gets called when you are writing the same record to the table. In this case, I want to ensure that you can not write an email in the users table into the email table. I added some tests cases for this and they seem to work as expected.

File: fxa-auth-db-server/index.js [hunk 12]
```diff
@@ -69,10 +69,23 @@ function createServer(db) {
...
+    op(function (req) {
+      return db.verifyEmail(req.params.id, req.params.emailCode)
+    })
+  )
   api.post('/account/:id/locale', withIdAndBody(db.updateLocale))
   api.get('/account/:id/sessions', withIdAndBody(db.sessions))
 
+  api.get('/account/:id/emails', withIdAndBody(db.accountEmails))
+  api.put('/account/:id/emails', withIdAndBody(db.createEmail))
+  api.post('/account/:id/emails', withIdAndBody(db.createEmail))
```
Comment (vbudhram) [inline fxa-auth-db-server/index.js]: I'll switch to POST to be consistent with others.

File: lib/db/mem.js [hunk 13]
```diff
@@ -969,6 +989,55 @@ module.exports = function (log, error) {
...
+    // Check to see if this email exists
+    var emailExistsInAccounts = Object.keys(accounts).some(function (uid) {
+      if (accounts[uid].normalizedEmail === data.normalizedEmail) {
+        return true
+      }
+    })
+
+    if (emailExistsInAccounts || emails[data.normalizedEmail]) {
+      throw error.duplicate()
+    }
```
Comment (vbudhram) [inline lib/db/mem.js]: Good point, will update.

File: lib/db/schema/patch-045-046.sql [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    DECLARE EXIT HANDLER FOR SQLEXCEPTION
+        BEGIN
+        ROLLBACK;
+        RESIGNAL;
+    END;
+
+    START TRANSACTION;
+
+    -- Currently, can not add an email that is specified in the
+    -- accounts table, regardless of verification state.
```
Comment (vbudhram) [inline lib/db/schema/patch-045-046.sql]: > If it makes the code cleaner, I would not be opposed to duplicating the primary address both in the accounts table, and in this new emails table.

Do you mean inserting email in both tables? When would that happen, when the user attempts to add an additional email?

Ex. Flow
1) Checks to see if there is an email listed as primary in the emails table.
2) If not, inserts email on the account table into the emails table and sets it as primary for the account.
3) Inserts additional email and sets `isPrimary` to false.
4) All subsequent queries will only need to check the emails table to ensure uniqueness.

File: lib/db/schema/patch-045-046.sql (hunk 7 above)
Comment (vbudhram) [inline lib/db/schema/patch-045-046.sql]: Oops, it isn't, was using it to test procedures.

File: lib/db/schema/patch-045-046.sql (hunk 10 above)
Comment (vbudhram) [inline lib/db/schema/patch-045-046.sql]: Yea I am not sure why that happened, looked fine in Webstorm. I'll try a different one to clean it up.

File: lib/db/mysql.js (hunk 2 above)
Comment (rfk) [inline lib/db/mysql.js]: Indeed, I mis-understood what this code what about.

File: lib/db/schema/patch-045-046.sql (hunk 14 above)
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: > Do you mean inserting email in both tables?

Yes.  It might turn out *not* to be cleaner overall, but maybe worth a look:

* When you first create the account, write the `email` field on the `accounts` table *and* insert a row into the `emails` table with `isPrimary=true`.
* Somehow backfill the primary emails for existing accounts into this table; this will probably have to be on-demand as you describe above, whenever you discover that the user has no emails in the `emails` table, add their primary email in.

Hmm...now that I try to describe it in more words, that "somehow backfill" hand-wavery sounds like it might be more complicated than it's worth.  I'll certainly defer to your judgement on which way is simpler overall!

**Comment by vbudhram** [comment]
@philbooth @rfk @seanmonstar Updated from feedback, mind another r?

File: lib/db/schema/patch-045-046.sql [hunk 15]
```diff
@@ -0,0 +1,224 @@
+-- Add emails table and corresponding procedures to
+-- create, get and delete them.
+
+CREATE TABLE IF NOT EXISTS emails (
+  normalizedEmail VARCHAR(255) NOT NULL PRIMARY KEY,
```
Comment (seanmonstar) [inline lib/db/schema/patch-045-046.sql]: Since we're creating a table of emails, this seems like an excellent time to make the primary key something smaller, like an auto incrementing integer. We've talked about eventually having the `accounts` and now `emailBounces` tables hold a foreign key to an `emails` table, to improve lookup times, and reduce the number of times we have the literal email text duplicated through the database.

File: lib/db/schema/patch-045-046.sql [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail AND uid = inUid AND isPrimary = 0;
+
+    -- Check to see if no rows were deleted, if this is the case, check to see if user
+    -- is attempting to deleted the email in the account table.
+    SELECT ROW_COUNT() INTO @deletedCount;
+    IF @deletedCount = 0 THEN
+      SELECT COUNT(*) INTO @isPrimary FROM accounts WHERE normalizedEmail = inNormalizedEmail AND uid = inUid;
+
+      IF @isPrimary = 1 THEN
+        SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Can not delete a primary email address.';
```
Comment (seanmonstar) [inline lib/db/schema/patch-045-046.sql]: Does this also need a `SET MYSQL_ERRNO = 1062`? Or I suppose a different number since it's not creating a duplicate...

File: lib/db/schema/patch-045-046.sql [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        inEmailVerified,
+        inKA,
+        inWrapWrapKb,
+        inAuthSalt,
+        inVerifierVersion,
+        inVerifyHash,
+        inVerifierSetAt,
+        inCreatedAt,
+        inLocale
+    );
```
Comment (seanmonstar) [inline lib/db/schema/patch-045-046.sql]: So I noticed that with a new account, the email is stored in the account table like normal. That email address is considered the user's primary email. What is the `isPrimary` column for in the `emails` table? I don't think I saw a way for that to be set to true in this PR, did I?

File: lib/db/schema/patch-045-046.sql [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+  uid BINARY(16) NOT NULL,
+  emailCode BINARY(16) NOT NULL,
+  isVerified BOOLEAN NOT NULL DEFAULT FALSE,
+  isPrimary BOOLEAN NOT NULL DEFAULT FALSE,
+  verifiedAt BIGINT UNSIGNED,
+  createdAt BIGINT UNSIGNED NOT NULL,
+  INDEX `uid` (`uid`)
+) ENGINE=InnoDB;
+
+DROP procedure IF EXISTS `createEmail_1`;
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: Indeed we do not

File: lib/db/mysql.js [hunk 19]
```diff
@@ -752,6 +753,59 @@ module.exports = function (log, error) {
...
+      DELETE_EMAIL,
+      [
+        uid,
+        email
+      ]
+    )
+      .catch(function(err){
+        // Signal exception is triggered when an attempt to
+        // delete a primary email.
+        if (err.errno === ER_SIGNAL_EXCEPTION) {
```
Comment (rfk) [inline lib/db/mysql.js]: +1, it feels like we should be checking for a specific errror number somehow, rather than just "did an exception occur?"

File: lib/db/schema/patch-045-046.sql [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+CREATE TABLE IF NOT EXISTS emails (
+  normalizedEmail VARCHAR(255) NOT NULL PRIMARY KEY,
+  email VARCHAR(255) NOT NULL,
+  uid BINARY(16) NOT NULL,
+  emailCode BINARY(16) NOT NULL,
+  isVerified BOOLEAN NOT NULL DEFAULT FALSE,
+  isPrimary BOOLEAN NOT NULL DEFAULT FALSE,
+  verifiedAt BIGINT UNSIGNED,
+  createdAt BIGINT UNSIGNED NOT NULL,
+  INDEX `uid` (`uid`)
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: nit: please check the naming conventions used for indexes on other tables and use something consistent; it appears we favour something like `tableName_columnName_columnName` which would make this `INDEX `emails_uid` for consistency.

File: lib/db/schema/patch-045-046.sql (hunk 18 above)
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: In fact if that procedure does already exist, we want to error out loudly.

File: lib/db/schema/patch-045-046.sql [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+    IN `createdAt` BIGINT UNSIGNED
+)
+BEGIN
+    -- Currently, can not add an email that is specified in the
+    -- accounts table, regardless of verification state.
+    SET @emailExists = 0;
+    SELECT COUNT(*) INTO @emailExists FROM accounts a WHERE a.normalizedEmail = normalizedEmail;
+    IF @emailExists > 0 THEN
+        SIGNAL SQLSTATE '45000' SET MYSQL_ERRNO = 1062, MESSAGE_TEXT = 'Email already exists in accounts table.';
+    END IF;
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: This will need to be done inside a transaction, otherwise there's a race condition where the above check succeeds, but someone concurrently registers that email address before we run the `INSERT` below.

File: lib/db/schema/patch-045-046.sql [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+BEGIN
+    SET @deletedCount = 0;
+    SET @isPrimary = 0;
+
+    -- Don't delete any email that are considered primary emails.
+    -- Since this should be the most common scenario, we try it first
+    DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail AND uid = inUid AND isPrimary = 0;
+
+    -- Check to see if no rows were deleted, if this is the case, check to see if user
+    -- is attempting to deleted the email in the account table.
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: s/to deleted/to delete/

File: lib/db/schema/patch-045-046.sql [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+
+
+CREATE PROCEDURE `verifyEmail_4`(
+    IN `inUid` BINARY(16),
+    IN `inEmailCode` BINARY(16)
+)
+BEGIN
+    SET @updatedCount = 0;
+
+    UPDATE accounts SET emailVerified = true WHERE uid = inUid AND emailCode = inEmailCode;
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: If the `emailCode` is not found in either table, should this procedure error out?

File: lib/db/schema/patch-045-046.sql [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+    IN `inLocale` VARCHAR(255)
+)
+BEGIN
+    -- Check to see if the normalizedEmail exists in the emails table before creating a new user
+    -- with this email.
+    SET @emailExists = 0;
+    SELECT COUNT(*) INTO @emailExists FROM emails WHERE normalizedEmail = inNormalizedEmail;
+    IF @emailExists > 0 THEN
+        SIGNAL SQLSTATE '45000' SET MYSQL_ERRNO = 1062, MESSAGE_TEXT = 'Unable to create user, email used belongs to another user.';
+    END IF;
```
Comment (rfk) [inline lib/db/schema/patch-045-046.sql]: As above, this check will need to go inside a transaction in order to avoid race conditions.

File: lib/db/mysql.js [hunk 25]
```diff
@@ -752,6 +753,59 @@ module.exports = function (log, error) {
...
+        data.createdAt
+      ]
+    )
+  }
+
+  // Select : emails
+  // Values : uid = $1
+  var ACCOUNT_EMAILS = 'CALL accountEmails_1(?)'
+  MySql.prototype.accountEmails = function (uid) {
+    return this.readOneFromFirstResult(
```
Comment (vbudhram) [inline lib/db/mysql.js]: Yea, believe it is. The name is kind of confusing to what it is actually doing, but it is returning `result[0]` similar to [securityEvents](https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/mysql.js#L1087).

File: lib/db/mysql.js (hunk 19 above)
Comment (vbudhram) [inline lib/db/mysql.js]: Updated to trigger a custom mysql errno 2100, can not delete primary email.

File: lib/db/schema/patch-045-046.sql (hunk 17 above)
Comment (vbudhram) [inline lib/db/schema/patch-045-046.sql]: Yup you are correct. In theory, the `isPrimary` will be utilized when the user has the ability the change their email address, which would then use the email in the emails table.

File: docs/API.md [hunk 26]
```diff
@@ -215,6 +220,60 @@ Returns:
...
+
+    * `uid` - (Buffer16) the uid of the account to get emails for
+    * `data`:
+      * email - (string) email to create
+      * normalizedEmail - (string) same as above but `.toLowerCase()`
+      * emailCode - (Buffer16) email code
+      * uid - (Buffer16) uid of the user
+      * isVerified - (number) 0|1 flag for whether or not email is verified
+      * isPrimary - (number) 0|1 flag for whether or not email is primary email
+      * createdAt - (number) date account was created with `Date.now()`
```
Comment (vbudhram) [inline docs/API.md]: Removed option to specify `createdAt`, uses `Date.now()`

**Comment by vbudhram** [comment]
@seanmonstar @rfk Updated from review, final r?

**Comment by vbudhram** [comment]
Thanks for the reviews guys!
